libyul/optimiser/VariableAssignmentMap.h

```diff
@@ -0,0 +1,83 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libyul/YulString.h>
+
+#include <set>
+#include <unordered_map>
+
+namespace solidity::yul
+{
+
+/**
+ * Class that implements a reverse lookup for an ``unordered_map<YulString, set<YulString>>`` by wrapping the
+ * two such maps - one ordered, and one reversed, e.g.
+ *
+ *  m_ordered           m_reversed
+ *  f -> (g,)           g -> (f,)
+ *  c -> (b,d,e,)       d -> (c,)
+ *  a -> (b,c,)         c -> (a,)
+ *                      e -> (c,)
+ *                      b -> (a,c,)
+ *
+ * The above example will from here onwards be referenced as ```Ref 1```.
+ *
+ * This allows us to have simultaneously managed insertion and deletion via a single interface, instead of manually
+ * managing this at the point of usage (see ``DataFlowAnalyzer``).
+ */
+class VariableAssignmentMap
+{
+public:
+	VariableAssignmentMap() = default;
+	/**
+	 * Insert a set of values for the provided key ``_variable`` into ``m_ordered`` and ``m_reversed``.
```

Comment:
What happens if `_variable` is already there?

Reply:
I guess [this](https://github.com/ethereum/solidity/pull/14112/files#r1175132871) answers the questions. But now I'm wondering - in case of the following assignments:

```
a = b
a = c
```
what should the references map look like?
```
a -> {b, c}
```
or just
```
a -> {c}
```
I guess we don't care about `b` at this point, since it's overwritten (at least in terms of `a`'s value) by `c`, so I assume we only care about the latest assignment to a variable?